### PR TITLE
add the interactive path and dendrogram

### PR DESCRIPTION
**Pull request by jasminezhuoy**
add the interactive path and dendrogram

### CODE REVIEW

File: R/plot_carp.R [hunk 1]
```diff
@@ -514,3 +426,497 @@ carp_heatmaply <- function(x,
...
+      animation_slider(currentvalue = list(prefix = "Regularization: ",suffix = "%")) #%>%
+  }
+}
+
+#' @noRd
+#' @importFrom rlang .data
+#' @importFrom dplyr filter select summarize pull
+#' @importFrom stats as.dendrogram
+#' @importFrom dendextend get_nodes_xy
+#' @importFrom plotly add_segments add_maekers add_text
```
Comment (michaelweylandt) [inline R/plot_carp.R:653]: Here and on L790, `add_maekers` should be `add_markers` (causing CI to fail)

File: DESCRIPTION [hunk 2]
```diff
@@ -53,3 +54,4 @@ Suggests: testthat,
 VignetteBuilder: knitr
 BugReports: https://github.com/DataSlingers/clustRviz/issues
 URL: https://github.io/DataSlingers/clustRviz, https://github.com/DataSlingers/clustRviz
+RoxygenNote: 6.1.1
```
Comment (michaelweylandt) [inline DESCRIPTION:57]: This one line shouldn't be included in this PR: it's autogenerated stuff

File: DESCRIPTION [hunk 3]
```diff
@@ -41,7 +41,8 @@ Imports:
     ggrepel,
     RColorBrewer,
     heatmaply,
-    gganimate
+    gganimate,
+    plotly
```
Comment (michaelweylandt) [inline DESCRIPTION:45]: Are we still using `shiny` after this PR? (I know you haven't touched heatmap code yet).

File: R/plot_carp.R [hunk 4]
```diff
@@ -82,160 +78,70 @@
 #' }
 plot.CARP <- function(x,
                       ...,
-                      type = c("dendrogram", "path", "dynamic_path", "js", "interactive"),
+                      type = c("dendrogram", "path", "js"),
```
Comment (michaelweylandt) [inline R/plot_carp.R:81]: Type `js` seems a bit uninformative since `interactive = TRUE` seems to imply javascript. Would `heatmap` be better?

File: R/plot_carp.R [hunk 5]
```diff
@@ -82,160 +78,70 @@
...
     dendrogram = {
-      carp_dendro_plot(x,
-                       percent = percent,
-                       k = k,
-                       dend.branch.width = dend.branch.width,
-                       dend.labels.cex = dend.labels.cex,
-                       dend.ylab.cex = dend.ylab.cex,
-                       ...)
+      if (interactive == FALSE) {
+        if (dynamic == FALSE) {
```
Comment (michaelweylandt) [inline R/plot_carp.R:98]: What happens here if `interactive = FALSE` but `dynamic = TRUE`? It looks like the function just peters out.

File: R/plot_carp.R [hunk 6]
```diff
@@ -514,3 +420,497 @@ carp_heatmaply <- function(x,
...
+
+  if(n_args > 1L){
+    crv_error("At most one of ", sQuote("percent"), " and ", sQuote("k"), " may be supplied.")
+  }
+
+  d <- as.dendrogram(x)
+  # get x/y locations of every node in the tree
+  m <- dendextend::get_nodes_xy(d)
+  colnames(m) <- c("x", "y")
+  allXY <- tibble::as_tibble(m)
```
Comment (michaelweylandt) [inline R/plot_carp.R:672]: This shouldn't need to be qualified with `tibble::` if the NAMESPACE is working properly.

File: R/plot_carp.R [hunk 7]
```diff
@@ -514,3 +420,497 @@ carp_heatmaply <- function(x,
...
+  has_k       <- !missing(k)
+  n_args      <- has_percent + has_k
+
+  if(n_args > 1L){
+    crv_error("At most one of ", sQuote("percent"), " and ", sQuote("k"), " may be supplied.")
+  }
+
+  d <- as.dendrogram(x)
+  # get x/y locations of every node in the tree
+  m <- dendextend::get_nodes_xy(d)
```
Comment (michaelweylandt) [inline R/plot_carp.R:670]: This shouldn't need to be qualified with `dendextend` if the NAMESPACE is working properly .

File: R/plot_carp.R [hunk 8]
```diff
@@ -514,3 +420,497 @@ carp_heatmaply <- function(x,
...
+  # but that isn't possible, right?
+  for (i in seq_along(heights)) {
+    allXY$label[[which(allXY$y == heights[i])]] <- labs[[i]]
+  }
+
+  tidy_segments <- dendextend::as.ggdend(d)$segments
+
+  allTXT <- allXY[allXY$y == 0, ]
+
+  allXY$members <- sapply(allXY$label, length)
```
Comment (michaelweylandt) [inline R/plot_carp.R:707]: I don't love `sapply` - can we use `vapply` here and elsewhere?

File: R/plot_carp.R [hunk 9]
```diff
@@ -514,3 +420,497 @@ carp_heatmaply <- function(x,
...
+  heights <- sapply(nodes, function(x) attr(x, "height"))
+  labs <- lapply(nodes, labels)
+
+  # NOTE: this won't support nodes that have the same height
+  # but that isn't possible, right?
+  for (i in seq_along(heights)) {
+    allXY$label[[which(allXY$y == heights[i])]] <- labs[[i]]
+  }
+
+  tidy_segments <- dendextend::as.ggdend(d)$segments
```
Comment (michaelweylandt) [inline R/plot_carp.R:703]: This shouldn't need the `dendextend` qualifier.

File: R/plot_carp.R [hunk 10]
```diff
@@ -514,3 +420,497 @@ carp_heatmaply <- function(x,
...
+
+  allTXT <- allXY[allXY$y == 0, ]
+
+  allXY$members <- sapply(allXY$label, length)
+  allTXT$label <- as.character(allTXT$label)
+
+  axis_x <- list( showgrid = F,
+                  zeroline = F,
+                  tickmode = "array",
+                  tickvals = 1:length(allTXT$label),
```
Comment (michaelweylandt) [inline R/plot_carp.R:713]: `seq_along` is more robust than `1:length`

File: R/plot_carp.R (hunk 5 above)
Comment (jasminezhuoy) [reply R/plot_carp.R:98]: Now it will show nothing since we don't have the dynamic dendrogram in previous version. We can add this in the future.

Comment (michaelweylandt) [reply R/plot_carp.R:98]: Can we put in a message / error (maybe the `.NotYetImplemented` function would help)? It seems like a bad idea to just do nothing silently.

File: R/plot_cbass.R [hunk 11]
```diff
@@ -1016,10 +1034,12 @@ cbass_dendro_plotly <- function(x,
       allTXT_dynamic <- rbind(allTXT_dynamic, allTXT_per)
     }
 
+    count <- NULL # for CMD check
```
Comment (michaelweylandt) [inline R/plot_cbass.R:1037]: This trick isn't great - if this is for use in a "tidyverse" function, it's better to change things to use the `.data` pronoun. 

This code could become

```
c <- dynamic_seg %>% group_by(.data$x, .data$xend) %>% summarize(count = n())
dynamic_sec <- arrange(merge(dynamic_seg, c), .data$Regularization, desc(.data$count), .data$x, .data$y)
```

I think you can also make this a bit cleaner with the `dplyr::count` function.

Comment (michaelweylandt) [reply R/plot_cbass.R:1037]: In general, the global variables thing isn't great because it's just "tricking" the check while the `.data` thing will actually make sure there's a real variable by that name. If you set `x <- NULL` and then try `select(data, x)` on `data` without an `x` column, you get a really nasty error.

File: R/plot_carp.R (hunk 4 above)
Comment (jasminezhuoy) [reply R/plot_carp.R]: Change `js` to `heatmap` would contribute to CMD check failure in `test_check("clustRviz", filter="carp_plot_static")` with `Message: 'arg' should be one of "dendrogram", "path", "heatmap"`

Comment (michaelweylandt) [reply R/plot_carp.R]: Yeah - the tests will need to be changed as well. 

See https://github.com/DataSlingers/clustRviz/blob/452701543af8874edb52a5ca5907b5ac3ebeb6ef/tests/testthat/test_carp_plot_static.R#L64 for this line (I think)

Comment (jasminezhuoy) [reply R/plot_carp.R]: I think we also need to add test_carp_plot_interactive.R or something like that to test added interactive plots.

Comment (michaelweylandt) [reply R/plot_carp.R]: Probably - but we should rethink these tests altogether at the end of the summer. Right now they don't do much other than just check they can run (they don't check that they are "right" in any sense) so I don't think it's such a big deal to have them in one file. Feel free to change the name if you think that makes sense.

File: R/plot_carp.R [hunk 12]
```diff
@@ -82,160 +83,76 @@
...
+        if (dynamic == FALSE) {
+          carp_dendro_plot(x,
+                           percent = percent,
+                           k = k,
+                           dend.branch.width = dend.branch.width,
+                           dend.labels.cex = dend.labels.cex,
+                           dend.ylab.cex = dend.ylab.cex,
+                           ...)
+        } else {
+          crv_error("Not implemented.")
```
Comment (michaelweylandt) [inline R/plot_carp.R:112]: Can you be a bit more specific in this error message (saying what isn't implemented.)

File: R/plot_carp.R [hunk 13]
```diff
@@ -82,160 +83,76 @@
...
-      if ( length(dots) != 0 ){
-        crv_error("Unknown arguments passed to ", sQuote("plot.CARP."))
+    heatmap = {
+      if (interactive==TRUE & dynamic==FALSE){
+        carp_heatmaply(x,
+                       ...,
+                       percent = percent,
+                       k = k)
+      } else {
+        crv_error("Not implemented.")
```
Comment (michaelweylandt) [inline R/plot_carp.R:154]: More details here.

File: R/plot_carp.R [hunk 14]
```diff
@@ -82,160 +83,76 @@
...
+                         percent.seq = percent.seq,
+                         ...)
+      }
     },
-    interactive = {
-      dots <- list(...)
-      if ( length(dots) != 0 ){
-        crv_error("Unknown arguments passed to ", sQuote("plot.CARP."))
+    heatmap = {
+      if (interactive==TRUE & dynamic==FALSE){
```
Comment (michaelweylandt) [inline R/plot_carp.R:148]: `if(interatctive & !dynamic)` should work here.

File: R/plot_carp.R [hunk 15]
```diff
@@ -514,3 +431,497 @@ carp_heatmaply <- function(x,
...
+
+  plot_frame_interactive <- bind_rows(lapply(percent.seq, function(pct){
+    ## Make a list of things to plot at each "percent" and then combine
+    plot_frame_full %>% filter(.data$GammaPercent <= pct & .data$GammaPercent > pct-0.01) %>%
+      mutate(Regularization = pct)
+  }))
+
+  mytext <- paste(plot_frame_init$ObsLabel)
+
+  if (dynamic == FALSE){
```
Comment (michaelweylandt) [inline R/plot_carp.R:525]: `if(!dynamic)` might be easier

**Comment by michaelweylandt** [comment]
@jasminezhuoy Something else to look at - for some reason, this PR has very low coverage. (See the codecov checks at the bottom of the PR - the ones with the pink umbrella). I know we can't test everything for graphics yet, but could you add some "smoke tests" (just running the code and making sure there are no errors) to the test infrastructure.

File: DESCRIPTION (hunk 3 above)
Comment (jasminezhuoy) [reply DESCRIPTION:45]: We don't use shiny anymore. I'll delete codes regrading shiny.
